// drizzle-orm-sqlite@0.16.1/db.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseSQLiteDatabase = void 0;
const drizzle_orm_1 = require("drizzle-orm");
const query_builders_1 = require("./query-builders");
const utils_1 = require("./utils");
class BaseSQLiteDatabase {
    constructor(
    /** @internal */
    dialect, 
    /** @internal */
    session) {
        this.dialect = dialect;
        this.session = session;
    }
    select(from) {
        const fields = (0, utils_1.orderSelectedFields)(from[drizzle_orm_1.Table.Symbol.Columns]);
        return new query_builders_1.SQLiteSelect(from, fields, this.session, this.dialect);
    }
    update(table) {
        return new query_builders_1.SQLiteUpdateBuilder(table, this.session, this.dialect);
    }
    insert(into) {
        return new query_builders_1.SQLiteInsertBuilder(into, this.session, this.dialect);
    }
    delete(from) {
        return new query_builders_1.SQLiteDelete(from, this.session, this.dialect);
    }
    run(query) {
        return this.session.run(query.getSQL());
    }
    all(query) {
        return this.session.all(query.getSQL());
    }
    get(query) {
        return this.session.get(query.getSQL());
    }
    values(query) {
        return this.session.values(query.getSQL());
    }
}
exports.BaseSQLiteDatabase = BaseSQLiteDatabase;
