// drizzle-orm-sqlite@0.16.1/query-builders/delete.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLiteDelete = void 0;
const table_1 = require("../table");
const utils_1 = require("../utils");
class SQLiteDelete {
    constructor(table, session, dialect) {
        this.table = table;
        this.session = session;
        this.dialect = dialect;
        this.run = (placeholderValues) => {
            return this.prepare().run(placeholderValues);
        };
        this.all = (placeholderValues) => {
            return this.prepare().all(placeholderValues);
        };
        this.get = (placeholderValues) => {
            return this.prepare().get(placeholderValues);
        };
        this.values = (placeholderValues) => {
            return this.prepare().values(placeholderValues);
        };
        this.config = { table };
    }
    where(where) {
        this.config.where = where;
        return this;
    }
    returning(fields = this.table[table_1.SQLiteTable.Symbol.Columns]) {
        this.config.returning = (0, utils_1.orderSelectedFields)(fields);
        return this;
    }
    /** @internal */
    getSQL() {
        return this.dialect.buildDeleteQuery(this.config);
    }
    toSQL() {
        return this.dialect.sqlToQuery(this.getSQL());
    }
    prepare() {
        return this.session.prepareQuery(this.toSQL(), this.config.returning);
    }
}
exports.SQLiteDelete = SQLiteDelete;
