// drizzle-orm-sqlite@0.16.1/query-builders/update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLiteUpdate = exports.SQLiteUpdateBuilder = void 0;
const table_1 = require("../table");
const utils_1 = require("../utils");
class SQLiteUpdateBuilder {
    constructor(table, session, dialect) {
        this.table = table;
        this.session = session;
        this.dialect = dialect;
    }
    set(values) {
        return new SQLiteUpdate(this.table, (0, utils_1.mapUpdateSet)(this.table, values), this.session, this.dialect);
    }
}
exports.SQLiteUpdateBuilder = SQLiteUpdateBuilder;
class SQLiteUpdate {
    constructor(table, set, session, dialect) {
        this.session = session;
        this.dialect = dialect;
        this.run = (placeholderValues) => {
            return this.prepare().run(placeholderValues);
        };
        this.all = (placeholderValues) => {
            return this.prepare().all(placeholderValues);
        };
        this.get = (placeholderValues) => {
            return this.prepare().get(placeholderValues);
        };
        this.values = (placeholderValues) => {
            return this.prepare().values(placeholderValues);
        };
        this.config = { set, table };
    }
    where(where) {
        this.config.where = where;
        return this;
    }
    returning(fields = this.config.table[table_1.SQLiteTable.Symbol.Columns]) {
        this.config.returning = (0, utils_1.orderSelectedFields)(fields);
        return this;
    }
    /** @internal */
    getSQL() {
        return this.dialect.buildUpdateQuery(this.config);
    }
    toSQL() {
        return this.dialect.sqlToQuery(this.getSQL());
    }
    prepare() {
        return this.session.prepareQuery(this.toSQL(), this.config.returning);
    }
}
exports.SQLiteUpdate = SQLiteUpdate;
