// drizzle-orm-sqlite@0.16.1/bun/session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreparedQuery = exports.SQLiteBunSession = void 0;
const drizzle_orm_1 = require("drizzle-orm");
const sql_1 = require("drizzle-orm/sql");
const utils_1 = require("drizzle-orm/utils");
const session_1 = require("../session");
class SQLiteBunSession extends session_1.SQLiteSession {
    constructor(client, dialect, options = {}) {
        var _a;
        super(dialect);
        this.client = client;
        this.logger = (_a = options.logger) !== null && _a !== void 0 ? _a : new drizzle_orm_1.NoopLogger();
    }
    exec(query) {
        this.client.exec(query);
    }
    prepareQuery(query, fields) {
        const stmt = this.client.prepare(query.sql);
        return new PreparedQuery(stmt, query.sql, query.params, this.logger, fields);
    }
}
exports.SQLiteBunSession = SQLiteBunSession;
class PreparedQuery extends session_1.PreparedQuery {
    constructor(stmt, queryString, params, logger, fields) {
        super();
        this.stmt = stmt;
        this.queryString = queryString;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
    }
    run(placeholderValues) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        return this.stmt.run(...params);
    }
    all(placeholderValues) {
        const { fields } = this;
        if (fields) {
            return this.values(placeholderValues).map((row) => (0, utils_1.mapResultRow)(fields, row));
        }
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        return this.stmt.all(...params);
    }
    get(placeholderValues) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        const value = this.stmt.get(...params);
        const { fields } = this;
        if (!fields) {
            return value;
        }
        return (0, utils_1.mapResultRow)(fields, value);
    }
    values(placeholderValues) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        return this.stmt.values(...params);
    }
}
exports.PreparedQuery = PreparedQuery;
