// .ignored/drizzle-orm-sqlite/sql.js/session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreparedQuery = exports.SQLJsSession = void 0;
const drizzle_orm_1 = require("drizzle-orm");
const sql_1 = require("drizzle-orm/sql");
const utils_1 = require("drizzle-orm/utils");
const session_1 = require("../session");
class SQLJsSession extends session_1.SQLiteSession {
    constructor(client, dialect, options = {}) {
        var _a;
        super(dialect);
        this.client = client;
        this.logger = (_a = options.logger) !== null && _a !== void 0 ? _a : new drizzle_orm_1.NoopLogger();
    }
    exec(query) {
        this.client.exec(query);
    }
    prepareQuery(query, fields) {
        const stmt = this.client.prepare(query.sql);
        return new PreparedQuery(stmt, query.sql, query.params, this.logger, fields);
    }
    prepareOneTimeQuery(query, fields) {
        const stmt = this.client.prepare(query.sql);
        return new PreparedQuery(stmt, query.sql, query.params, this.logger, fields, true);
    }
}
exports.SQLJsSession = SQLJsSession;
class PreparedQuery extends session_1.PreparedQuery {
    constructor(stmt, queryString, params, logger, fields, isOneTimeQuery = false) {
        super();
        this.stmt = stmt;
        this.queryString = queryString;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
        this.isOneTimeQuery = isOneTimeQuery;
    }
    run(placeholderValues) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        const result = this.stmt.run(params);
        if (this.isOneTimeQuery) {
            this.free();
        }
        return result;
    }
    all(placeholderValues) {
        const { fields } = this;
        if (fields) {
            return this.values(placeholderValues).map((row) => (0, utils_1.mapResultRow)(fields, row.map(normalizeFieldValue)));
        }
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        this.stmt.bind(params);
        const rows = [];
        while (this.stmt.step()) {
            rows.push(this.stmt.getAsObject());
        }
        if (this.isOneTimeQuery) {
            this.free();
        }
        return rows;
    }
    get(placeholderValues) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        const { fields } = this;
        if (!fields) {
            const result = this.stmt.getAsObject(params);
            if (this.isOneTimeQuery) {
                this.free();
            }
            return result;
        }
        const row = this.stmt.get(params);
        if (this.isOneTimeQuery) {
            this.free();
        }
        return (0, utils_1.mapResultRow)(fields, row.map(normalizeFieldValue));
    }
    values(placeholderValues) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        this.stmt.bind(params);
        const rows = [];
        while (this.stmt.step()) {
            rows.push(this.stmt.get());
        }
        return rows;
    }
    free() {
        return this.stmt.free();
    }
}
exports.PreparedQuery = PreparedQuery;
function normalizeFieldValue(value) {
    if (value instanceof Uint8Array) {
        if (typeof Buffer !== 'undefined') {
            if (!(value instanceof Buffer)) {
                return Buffer.from(value);
            }
            return value;
        }
        if (typeof TextDecoder !== 'undefined') {
            return new TextDecoder().decode(value);
        }
        throw new Error('TextDecoder is not available. Please provide either Buffer or TextDecoder polyfill.');
    }
    return value;
}
